File: src/test/java/net/igsoft/sdi/testclasses/BCreator.java
package net.igsoft.sdi.testclasses;

import java.util.List;

import com.google.common.collect.Lists;

import net.igsoft.sdi.creator.CreatorBase;
import net.igsoft.sdi.engine.InstanceProvider;
import net.igsoft.sdi.parameter.LaunchType;

public class BCreator extends CreatorBase<BClass, LaunchType> {
    @Override
    public BClass create(InstanceProvider instanceProvider, LaunchType launchType) {
        DClass d = instanceProvider.getOrCreate(DClass.class, launchType);
        PClass p = instanceProvider.getOrCreate(PClass.class, new PParametrizedCreator.Params(false, "id"));

        return new BClass(d, p, instanceProvider.getOrCreate(Stepper.class));
    }

    @Override
    public List<CreatorBase<?, ?>> defaultCreators() {
        return Lists.newArrayList(new PParametrizedCreator());
    }
}

File: src/test/scala/net/igsoft/sdi/scala/testclasses/BCreator.scala
package net.igsoft.sdi.scala.testclasses

import net.igsoft.sdi.{Creator, InstanceCreator, LaunchType}

class BCreator extends Creator[B, LaunchType] {
  override def create(instanceCreator: InstanceCreator, launchType: LaunchType): B = {
    val d = instanceCreator.getOrCreate(classOf[D], launchType)
    val p = instanceCreator.getOrCreate(classOf[P], new PCreatorParams(false, "id"))
    new B(d, p, instanceCreator.getOrCreate(classOf[Stepper]))
  }
}
